File: kmeans.c
#include <stdio.h>;
#include <stdlib.h>;
#include <assert.h>;
#include <kmeans.h>; //TODO: why is this here? Commenting out derives several errors in code


typedef enum HasChangedEnum {
    TRUE = 1,
    FALSE = 0
} HasChanged;


/*** Function Declaration ***/

int get_num_points(FILE *fptr);
int get_num_coordinates(FILE *fptr);
void get_points_from_file (int numPoints, int numCoordinates, FILE *fptr, double *points);
void reset_array_double(double * arr, int size);
void reset_array_int(int * arr, int size);
double* split_points_to_clusters(double* centroids, double* points, int point_size, int k, int points_num);
HasChanged calculate_new_centroids(double *centroids, double *new_centroids, int* centroids_count, int point_size, int k);
double get_distance (double *points, int startIndex1, int startIndex2, int numCoordinates);


/*** Main Function ***/

int main(int argc, char **argv) { 
    //argc = number of parameters (including program name), **argv = string array of parameters. Reference: https://stackoverflow.com/questions/29584898/c-getting-input-from-cmd
    //TODO: make sure how input is actually formatted
    assert (argc == 3 || argc = 4);
    int k, max_iter;
    char *filename;
    int numPoints, numCoordinates;

    if (argc == 3) { //max_iter not passed as argument
        assert (atoi(argv[0] > 0));
        k = atoi(argv[0]);
        max_iter = 200;
        filename = argv[2];
    }
    else { //argc == 4
        assert (atoi(argv[1] > 0 && atoi(argv[2] > 0)); 
        k = atoi(argv[0]);
        max_iter = atoi(argv[1]);
        filename = argv[4];
    }

    FILE *fptr;
    fptr = fopen(filename,"r");
    if (fptr != NULL) {
        numPoints = get_num_points(fptr);
        numCoordinates = get_num_coordinates(fptr);
    }

    double *points = malloc(sizeof *points * numPoints * numCoordinates); //TODO: probably not the best way to allocate memory for 2D-array. Reference: https://stackoverflow.com/questions/9381711/how-to-declare-a-variable-size-2d-array-in-c
    get_points_from_file(numPoints, numCoordinates, fptr, points);

    fclose(fptr);

    // initiate the centroids to be the first k points
    double *centroids = malloc(sizeof *centroids k * numPoints);
    for (int i=0; i<k; i++) {
        for (int j=0; j< numCoordinates; j++) {
            centroids[i*numCoordinates + j] = points[i*numCoordinates + j] 
        }
    }

    while (max_iter > 0) {
        split_points_to_clusters (centroids, points, numCoordinates, k, numPoints);
        if (calculate_new_centroids(centroids, , , k == FALSE)) { //TODO: what should "new_centroids" and "controid_count" parameters be? 
            max_iter = 0; //breaks the loop
        } 
        max_iter -= 1;
    }

    //print final centroids
    for (int i=0; i<k; i++) {
        int j = 0;
        while (j < numCoordinates -1) {
            printf ("%lf," , centroids[i*numCoordinates + j]);
            j++;
        }
        printf ("%lf/n" , centroids[i*numCoordinates + j]) //print last coordinate
    }

    return 0;

}


/*** Function Implementation ***/

int get_num_points(FILE *fptr) {
    int countPoints = 1;
    char ch;
    while (ch = fgetc(fptr) != EOF) {
        if (ch == '\n') {
            countPoints++;
        }
    }
    rewind(fptr);
    return countPoints;
}

int get_num_coordinates(FILE *fptr) {
    int countCoordinates = 1;
    char ch;
    while (ch = fgetc(fptr) != '\n') {
        if (ch == ',') {
            countCoordinates++;
        }
    }
    rewind(fptr);
    return countCoordinates;
}

void get_points_from_file (int numPoints, int numCoordinates, FILE *fptr, double *points) {
    for (int pointIndex=0; pointIndex<numPoints; pointIndex++) {
        for (int coordinateIndex=0; coordinateIndex<numCoordinates; coordinateIndex++) {
            fscanf(fptr , "%lf,\n", points[pointIndex * numCoordinates + coordinateIndex]); //Reference: https://stackoverflow.com/questions/43286609/how-to-use-fscanf-to-read-doubles-in-c 
            //TODO: make sure this fscanf format actually catches all of our doubles
        }
    }
    
}


void reset_array_double(double * arr, int size) {
    for (int i = 0; i < size; ++i) {
        *(arr + i) = 0;
    }
}

void reset_array_int(int * arr, int size) {
    for (int i = 0; i < size; ++i) {
        *(arr + i) = 0;
    }
}

double* split_points_to_clusters(double* centroids, double* points, int point_size, int k, int points_num) {
    double min_distance, distance;
    int min_centroid;
    double* new_centroids = (double *)calloc(sizeof(double ), point_size * k);
    reset_array_double(new_centroids, point_size * k)
    int* clusters_count = (int *)calloc(sizeof(int), k);
    reset_array_int(clusters_count, k)
    for (int i = 0; i < points_num; ++i) { // iterate points
        min_centroid = 0;
        min_distance = get_distance(centroids, points, 0, i);
        for (int j = 1; j < k; ++j) { // check min distance from all the centroids
            distance = get_distance(centroids, points, (j * point_size), i);
            if (distance < min_distance) {
                min_distance = distance;
                min_centroid = j;
            }
        }
        for (int l = 0; l < point_size; ++l) { // sum points
            *(new_centroids + (j * point_size) + l) += *(points + i + l);
        }
        *(clusters_count + min_centroid) ++; // increase counter
    }

    //TODO: what about return value?
}

HasChanged calculate_new_centroids(double *centroids, double *new_centroids, int* centroids_count, int point_size, int k) {
    HasChanged has_changed = FALSE
    for (int i = 0; i < k; ++i) { // calculate average
        for (int j = 0; j < point_size; ++j) {
            *(new_centroids + (i * point_size) + j) = *(new_centroids + (i * point_size) + j) / *(centroids_count + i);
            if (*(new_centroids + (i * point_size) + j) != *(centroids + (i * point_size) + j)) {
                has_changed = TRUE;
            }
        }
    }
    return has_changed;
}



double get_distance (double *points, int startIndex1, int startIndex2, int numCoordinates) { //TODO: adjust parameter to the function use in split_points_to_clusters
    double distance = 0;
    for (int i=0; i<numCoordinates; i++) {
        distance += pow(points[startIndex1 + i] - points[startIndex2 + i], 2);
    }
    return distance;
}

File: kmeans.py
import sys

POINTS_SEPARATOR = '\n'
COORDINATES_SEPARATOR = ','


def get_distance(p1, p2):
    assert len(p1) == len(p2)
    distance = 0
    for i in range(len(p1)):
        distance += (p1[i] - p2[i]) ** 2
    return distance


def get_points(filename):
    points = []
    with open(filename, 'r') as f:
        data = f.read()
    for point in data.split(POINTS_SEPARATOR):
        if len(point) != 0:
            points.append([float(n) for n in point.split(COORDINATES_SEPARATOR)])
    return points


def split_points_to_clusters(centroids, points):
    for c in centroids:
        c["sum"] = [0] * len(points[0])
        c["count"] = 0
    for point in points:
        min_cluster = 0
        min_distance = get_distance(point, centroids[0]["point"])
        for i in range(1, len(centroids)):
            distance = get_distance(point, centroids[i]["point"])
            if distance < min_distance:
                min_cluster, min_distance = i, distance
        centroids[min_cluster]["count"] += 1
        for j in range(len(point)):  # update sum of the point for the cluster
            centroids[min_cluster]["sum"][j] += point[j]


def calculate_new_centroids(centroids):
    """
    Updates the centroids and return if one of them has changed.
    """
    is_changed = False
    for i in range(len(centroids)):
        new_centroid = []
        for j in range(len(centroids[0]["point"])):
            new_centroid.append(centroids[i]["sum"][j] / centroids[i]["count"])  # calculate the average
            if new_centroid[j] != centroids[i]["point"][j]:
                is_changed = True
                centroids[i]["point"] = new_centroid
    return is_changed


def main():
    assert len(sys.argv) in [3, 4]
    k = int(sys.argv[1])
    filename = sys.argv[2]
    max_iter = int(sys.argv[3]) if len(sys.argv) > 3 else 200  # the default max_iter
    points = get_points(filename)
    centroids = []
    for i in range(k):  # initiate the centroids to be the first k points
        centroids.append({"point": points[i]})
    while max_iter > 0:
        split_points_to_clusters(centroids, points)
        if not calculate_new_centroids(centroids):  # there is no change in the centroids
            max_iter = 0  # breaks the loop
        max_iter -= 1
    output = POINTS_SEPARATOR.join([COORDINATES_SEPARATOR.join([str(round(i, 4)) for i in c["point"]]) for c in centroids])
    print(output)
    return output


if __name__ == '__main__':
    main()
